### PR TITLE
allow IPv6 service subnet to be /64

### PR DESCRIPTION
**Pull request by danwinship**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Almost all "normal" IPv6 subnets are /64 so it's weird to have the IPv6 service subnet be required to be a /108 or greater. The fact that we don't support allocating 2^64 service IPs isn't really relevant; an Ethernet segment cannot plausibly support 2^64 NICs, but IPv6-over-Ethernet is mostly only defined for /64 subnets.

The service allocator code already ignores all but the low 16 bits of the subnet for IPv6 anyway, so this is mostly just cosmetic: you used to say `"fd03::/108"` and it would actually allocate only out of
`"fd03::/112"`. Now you can say `"fd03::/64"` but it will still actually allocate only out of `"fd03::/112"`.

Also, reword the validation error messages to try to avoid ambiguity between larger/smaller subnet size vs larger/smaller prefix length (which are inverses).

**Which issue(s) this PR fixes**:
none, though it's an alternative to https://github.com/kubernetes/enhancements/pull/1534, which proposed changing the allocator code so we could actually allocate out of a full /64

**Does this PR introduce a user-facing change?**:
```release-note
In IPv6 and dual-stack clusters, the service CIDR can now be specified as
a /64 rather than needing to be /108 or /112. (The number of service IPs
that can actually be allocated is unchanged; this is just a cosmetic change.)
```

/sig network
/priority important-longterm

/assign @aojea

### CODE REVIEW

File: pkg/registry/core/service/ipallocator/allocator.go [hunk 1]
```diff
@@ -82,14 +82,28 @@ type Range struct {
 
 // NewAllocatorCIDRRange creates a Range over a net.IPNet, calling allocatorFactory to construct the backing store.
 func NewAllocatorCIDRRange(cidr *net.IPNet, allocatorFactory allocator.AllocatorFactory) (*Range, error) {
-	max := RangeSize(cidr)
-	base := bigForIP(cidr.IP)
+	max := utilnet.RangeSize(cidr)
```
Comment (aojea) [inline pkg/registry/core/service/ipallocator/allocator.go:85]: This rangesize is not capping to 16kb, max is 2^64

https://github.com/kubernetes/utils/blob/3d5a2fe318e4201932f44a9b6ee4ef0c6221f1d3/net/net.go#L168-L180

File: pkg/registry/core/service/ipallocator/allocator.go [hunk 2]
```diff
@@ -82,14 +82,28 @@ type Range struct {
...
+	base := utilnet.BigForIP(cidr.IP)
 	rangeSpec := cidr.String()
 
+	if utilnet.IsIPv6CIDR(cidr) {
+		// Limit the max size, since the allocator keeps a bitmap of that size.
+		if max > 65536 {
+			max = 65536
+		}
+	} else {
+		// Don't use the IPv4 network's broadcast address.
```
Comment (aojea) [inline pkg/registry/core/service/ipallocator/allocator.go:95]: you got me here and I had to check the book  :sweat_smile:  IPv6 does NOT have broadcast addresses XD

File: pkg/registry/core/service/ipallocator/allocator.go (hunk 1 above)
Comment (aojea) [reply pkg/registry/core/service/ipallocator/allocator.go:85]: ups, it's capped below , scratch it

**Comment by aojea** [comment]
yeah, I think that this is a good next step, and perfectly explained.
I think it will help IPv6 deployments simplifying the address planning, allowing people to use /64 for the service subnet.
It's LGTM but prefer to wait for @liggitt and @thockin opinions

**Comment by danwinship** [comment]
So one issue I realized belatedly: it would be reasonable for someone to assume that if they specify a service CIDR of `fd03::/64` (and kube accepts it) then that means they can manually assign a ClusterIP of something like `fd03::1:2:3:4` to a service if they want to. But that would fail.

OTOH, the limit to 16 bits of addresses makes it pretty easy to explain: an IPv6 ClusterIP cannot have any `:`s after the `::`

**Comment by neolit123** [comment]
> The service allocator code already ignores all but the low 16 bits of the subnet for IPv6 anyway, so this is mostly just cosmetic: you used to say "fd03::/108" and it would actually allocate only out of
"fd03::/112". Now you can say "fd03::/64" but it will still actually allocate only out of "fd03::/112".

are we documenting the existing v6 behavior somewhere?
i personally find what k8s is doing before and after this change to be a bit strange.

> fd03::/64

the users would expect this to give them exp2(64) hosts, even if they would not need that many.
my vote goes for bitmap compression and actual hosts numbers.

**Comment by danwinship** [comment]
I guess another possibility would be that if you specify a /64, then instead of keeping a bitmap we just allocate IPs randomly and assume that the 1 in 18,000,000,000,000,000,000 chance of duplicate IPs won't happen

**Comment by aojea** [comment]
I have second thoughts about this approach, Lubomir has a great point, 

> the users would expect this to give them exp2(64) hosts, even if they would not need that many.

you are right, perpetuating the workaround may not be the best solution, and it will cause a bad UX

> we just allocate IPs randomly and assume that the 1 in 18,000,000,000,000,000,000 chance of duplicate IPs won't happen

I think one of the biggest constraints with the current ipallocator is that we have to persist the data in etcd to make it available to multiple apiservers, which [also imposes some limits (1.5MB by default) on the serialization/deserialization size](https://github.com/etcd-io/etcd/blob/master/Documentation/dev-guide/limit.md#request-size-limit) 

> my vote goes for bitmap compression and actual hosts numbers.

I think that we should allow using whatever IP that is available in the defined range, but that we can limit the number of IPs, i.e., don't think that having more than 1 million of IPs make sense (at this time :smile: ) , there are other components that will crash before IMHO.

I have a prototype with roaring bitmaps working, but they are not convincing me, the compression is not all magic and there are still limits (2^32) because of the index they use.

Let me share my thoughts and my experiments with other [alternatives I've tried](https://github.com/aojea/go-radix/tree/ipradix) ... Using a map[string]bool or a radix tree the problem is the overhead and the serialization/deserialization. 

I've added some benchmarks with 1M IP addresses, I'm not an expert on go benchmarking so I think that there are a lot of mistakes, but it may be useful to have an idea

```
BenchmarkInsertIPv4-8      	  541966	      1995 ns/op	     218 B/op	       5 allocs/op
BenchmarkInsertIPv6-8      	  622402	      2944 ns/op	     217 B/op	       4 allocs/op
BenchmarkInsertMapIPv4-8   	 1000000	      1163 ns/op	      98 B/op	       2 allocs/op
BenchmarkInsertMapIPv6-8   	  887077	      1373 ns/op	     134 B/op	       2 allocs/op
BenchmarkLookupIPv4-8      	       1	2213150108 ns/op	194624880 B/op	 4391922 allocs/op
BenchmarkLookupIPv6-8      	       1	1925066770 ns/op	215001072 B/op	 4594806 allocs/op
BenchmarkLookupMapIPv4-8   	       1	1064116422 ns/op	98826464 B/op	 2038264 allocs/op
BenchmarkLookupMapIPv6-8   	       2	 541852708 ns/op	61395104 B/op	 1019019 allocs/op

```

1Million IPs
bitmap: 1bit per IP -> 12,5 Mbytes
map: 103B/op =>103 Mbytes
radix: 214B/op > 214 Mbytes

I also though in using etcd directly as a K/V but the interface has to implement:

* Allocate()
* Release()
* AllocateNext()
* Snapshot()
* Restore()

and AllocateNext() needs to find a free address, I think that iterating over all keys will add a lot of stress to etcd.

thoughts?

**Comment by danwinship** [comment]
> > we just allocate IPs randomly and assume that the 1 in 18,000,000,000,000,000,000 chance of duplicate IPs won't happen
> 
> I think one of the biggest constraints with the current ipallocator is that we have to persist the data in etcd to make it available to multiple apiservers

In the "allocate randomly" case, there is no data to persist. Assuming the random numbers are plausibly random, then one or multiple controllers, across one or more runs, can allocate randomly independently of each other.

I guess to deal with humans manually assigning ClusterIPs to some services in a not-especially-random way, we'd still want to sanity-check manually-assigned IPs. But since allocation of new IPs would still be primarily random, we don't need a fancy data structure that optimizes for being able to find an unused IP. We can just keep a `sets.Int64` or whatever, and initialize it from the set of Services on startup. Kube-proxy already keeps per-Service maps and fills them in from etcd on startup. There's no reason the service controller can't do it too...

(FTR though I'm less convinced that this PR is a good idea at this point.)

**Comment by neolit123** [comment]
> we just allocate IPs randomly and assume that the 1 in 18,000,000,000,000,000,000 chance of duplicate IPs won't happen

this works fine for lower leading bits in the CIDR, but will get progressively worse performance wise near `/128` as the PRNG will may face collisions. the lower the range - the higher the probability of collisions. still worth benchmarking this approach.

> In the "allocate randomly" case, there is no data to persist. Assuming the random numbers are plausibly random, then one or multiple controllers, across one or more runs, can allocate randomly independently of each other.

hm, how would k8s manage colliding IPs in that case?

> that means that the overhead is bigger with current implementation

for host allocation, the current implementation seems to be benchmarking better because it has no new memory allocations per operation. might be possible to optimize the radix in that regard.

in terms of lookup this has to be compared to the current implementation, but i'd assume the current one is much faster too.

> I also though about using etcd directly as a K/V but the interface has to implement:

the maintainers might not be happy about using etcd directly.
there is an accumulating opinion that longterm k8s should abstract away it's storage backend to better support low spec devices.

**Comment by danwinship** [comment]
> hm, how would k8s manage colliding IPs in that case?

How does k8s manage colliding `ObjectMeta.UID`s? (Spoiler alert: I'm pretty sure it doesn't. It definitely didn't used to.)

Except that, as I noted later, we'd probably have to check for collisions because of user-specified IPs. But that's not hard; we just build a hash table. We already have other per-Service hash tables, so this doesn't impose any new scaling limit. "Random assignment followed by hash-table-based sanity-check" is less efficient than the current approach in the IPv4 case because when the service CIDR is close to full it's harder to find an unused IP by poking around at random. But in the case of a /64, the CIDR will never be close to full.

**Comment by neolit123** [comment]
> How does k8s manage colliding ObjectMeta.UIDs? (Spoiler alert: I'm pretty sure it doesn't. It definitely didn't used to.)

that is of course correct:
https://github.com/google/uuid/blob/bd451584982ecf4ca5b1e5938cf168e17e30d837/version4.go#L17-L31

even if the docs say:
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids
> Every object created over the whole lifetime of a Kubernetes cluster has a distinct UID

but with these UUID there is the same large constant pool of N, while for a CIDR we have to account for small ranges, where collisions can happen even for system generated hosts.

> Except that, as I noted later, we'd probably have to check for collisions because of user-specified IPs. But that's not hard; we just build a hash table. 

hash both system generated and user generated hosts?

**Comment by danwinship** [comment]
> but with these UUID there is the same large constant pool of N, while for a CIDR we have to account for small ranges, where collisions can happen even for system generated hosts.

right, we'd definitely want to keep the old code for "small" service CIDRs.

> hash both system generated and user generated hosts?

Yes. (In particular, after a Service has been created there's no indication of whether its ClusterIP was system-generated or user-generated, so if you restart the service controller it would no longer know which ones were which.)

**Comment by neolit123** [comment]
> right, we'd definitely want to keep the old code for "small" service CIDRs.

last time i checked the largest number of hosts we currently support for v4 and v6 was 2^16.
does this mean that for more than 2^16 we are going to pick random non-colliding hosts?

the chances for collision above but near the 2^16 range are relatively high.

**Comment by danwinship** [comment]
For IPv4 we currently allow anything from /12 to /29. For IPv6 we currently allow anything from /108 to /125 except that /108, /109, /110, and /111 are all silently treated as though they were /112.

My suggestion was that for any subnet allowed by the old code, we continue using the old code, but that we also allow /64s for IPv6. But *only* /64s. No /48s or /72s or /107s. All of the documentation would be updated to use /64s, but we would just continue to accept /108-/125 for backward compat, and in that case we'd continue to use the old code.

**Comment by aojea** [comment]
I'm still thinking that we can support everything with one node path, we should leverage  that is impossible to use the whole range, I would say that more than 100K IPs sound a good upper limit, let's say 1 Million IPs, that should be ok for a map[int64]bool

I've submitted this PR with that approach but replacing the cidr_sets bitmaps with a map instead of the ip allocator, because they don't need to persist the data in etcd https://github.com/kubernetes/kubernetes/pull/90184


the `map[int64]bool` seems promising, checking the benchmarks it scales for 2^62 ranges, the bitmap approach stops to answer in less than a minute for  2^16

```
BenchmarkAllocateNext_48_52-4            2153733           658 ns/op           0 B/op          0 allocs/op
BenchmarkAllocateNext_48_56-4             207295          8037 ns/op           0 B/op          0 allocs/op
BenchmarkAllocateNext_48_60-4              19568         70644 ns/op          18 B/op          0 allocs/op
BenchmarkAllocateNextMap_48_52-4        24808773            51.2 ns/op         0 B/op          0 allocs/op
BenchmarkAllocateNextMap_48_56-4        24824025            53.7 ns/op         0 B/op          0 allocs/op
BenchmarkAllocateNextMap_48_60-4        22987214            49.8 ns/op         0 B/op          0 allocs/op
BenchmarkAllocateNextMap_48_64-4        21648500            55.6 ns/op         0 B/op          0 allocs/op
BenchmarkAllocateNextMap_48_88-4         1000000          1024 ns/op         270 B/op          7 allocs/op
BenchmarkAllocateNextMap_48_100-4        1000000          1054 ns/op         270 B/op          7 allocs/op
```

**Comment by danwinship** [comment]
> I'm still thinking that we can support everything with one code path, we should leverage that is impossible to use the whole range

But it *is* possible to use the whole range for IPv4 clusters. Eg, you could have a /22 CIDR with 1000 services (so there are only 24 free IPs scattered through the range)

**Comment by aojea** [comment]
Just an update because I've found a problem with the benchmarks. I'm totally sorry because I was following a red herring, the range allocator performance problem I was describing is not related to the service ip allocator.

> the `map[int64]bool` seems promising, checking the benchmarks it scales for 2^62 ranges, the bitmap approach stops to answer in less than a minute for 2^16

I did a more basic test to compare a map against a bitmap, I've created and allocated all addresses in a map or bitmap with 1 Million and 100K IPs, took a snapshot and checked its length (is a slice of  bytes) , measuring each of the operations.
You can see the results here https://github.com/kubernetes/kubernetes/pull/90213#issuecomment-615198589 
The size of the snapshots using maps are too big for an allocator.
Again, sorry for the noise, :bow:

**Comment by lavalamp** [comment]
There is no need to make our data structure express that many IPs. It's not realistic for a single cluster.

(Aside: now that we have webhooks, it's actually very wrong for apiserver to keep this private data. This should be implemented via webhook.)

I would limit the bitmap to 2^20 IPs for now, no matter how big the range is. What is the harm in not allocating some of the IPs in the range?

Making a sparse bitmap type thing just doesn't seem worth it, and randomized selection doesn't seem safe, although if you know of a birthday problem calculator that'll take 2^64 as an input we can check that.

(sorry, I wrote this yesterday and forgot to push "send")

**Comment by aojea** [comment]
> What is the harm in not allocating some of the IPs in the range?

There is no harm, just that some users may want to allocate static IPs, and this will avoid that, if you have a /64, that's the usual subnet size for an IPv6 subnet , it sounds like a lot of voids :)

> (Aside: now that we have webhooks, it's actually very wrong for apiserver to keep this private data. This should be implemented via webhook.)

@lavalamp Sorry if it is a dumb question, the webhook will solve some of the allocation problems or is just that "is the right place for the ipallocator"?

**Comment by lavalamp** [comment]
> There is no harm, just that some users may want to allocate static IPs, and this will avoid that, if you have a /64, that's the usual subnet size for an IPv6 subnet , it sounds like a lot of voids :)

Ah right, the problem is service authors choosing IPs, not the allocator. doh.

Here's an idea:
* Make the bit array a fixed size.
* If the CIDR is larger than that fixed size, set the bit (IP number) % (bit array size).
* This means IPs with the same number % size will collide. That will be confusing for users, but a good error message will help.
* A tactically chosen table size (prime, at least relatively with 2 and 10) may make that message less rare.

I may think of other ideas later.


> Sorry if it is a dumb question, the webhook will solve some of the allocation problems or is just that "is the right place for the ipallocator"?

The latter; it would mean that the ip space could be divvied up and tracked however the authors find it convenient. There is no good reason for kube-apiserver to be opinionated about this now; in the early days, we didn't have a choice.

**Comment by lavalamp** [comment]
Another strategy might be to declare that if the CIDR gives more than 2^N IPs, then we require that for every IP, the bottom M bits be zero, such that 2^(N-M) is a reasonable number of IPs. This is exactly the same as using only the first 2^M IPs, except that the zero bits are at the end rather than in the middle of the IP address. The only advantage of doing this is that if a user tries to occupy a not-tracked IP address, we can suggest a nearby one, since valid IPs will be distributed uniformly rather than all at the beginning of the range.

**Comment by danwinship** [comment]
With IPv6 you'll get shorter/cleaner addresses if you use the bottom of the range, because you can only use `::` to eliminate extra 0s once. So if you're using `fd03::/64` as the prefix, then allocating at the bottom of the range you'd have IPs like `fd03::1234`, while allocating at the top of the range you'd have IPs like `fd03::1234:0:0:0`.


Another possibility, if we can assume that most IPs are going to be system-allocated, not user-assigned, would be to use the bitmap for IPs in the system-allocated range, and a different method for recording IPs outside that range.

**Comment by lavalamp** [comment]
> then allocating at the bottom of the range you'd have IPs like fd03::1234, while allocating at the top of the range you'd have IPs like fd03::1234:0:0:0.

I guess I was assuming that the `::` likely was used already in the base address, since how likely is it that the service CIDR would be precisely the first thing in that range? But if that's more likely than I think, then I agree with you point.

If we wanted to permit users to select IPs even if we only allocate from a fixed block at the beginning, I'd expect some variant of a bloom filter?

* They can be deterministic if the set size is known (in our case, it is): https://en.wikipedia.org/wiki/Bloom_filter#Avoiding_false_positives_in_a_finite_universe
* counted (for permitting deletions): https://en.wikipedia.org/wiki/Counting_Bloom_filter

If we use a bloom filter, instead of deleting entries we can just occasionally reconstruct it. This forbids immediate reuse of an IP address which is a feature most of the time but would be really annoying if you wanted to e.g. recreate a service object and reuse an IP.

> [random selection] and assume that the 1 in 18,000,000,000,000,000,000 chance of duplicate IPs won't happen

(from much earlier in the thread) This is sadly not how it works, since it's repeated. 1/billion is my personal threshold. I just made a spreadsheet with a [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem) approximation, and most big CIDRs are not big enough.

birthday64

**Comment by danwinship** [comment]
> I guess I was assuming that the `::` likely was used already in the base address, since how likely is it that the service CIDR would be precisely the first thing in that range? But if that's more likely than I think, then I agree with you point.

For "real" IPv6 networks, yes, the prefix will probably fill out the entire top half of the address. But for the service CIDR, the IPv6 private address space is `fd00::/8` which gives you 256 private subnets to work with even if you only use the `fdXY:0:0:0::` ones, so probably most people aren't going to use something longer than that.

**Comment by agowa338** [comment]
Just one thought regarding the implementation of AllocateNext.
Does it need to be deterministic? E.g. do we need it to do "i++", or would it also be fine to just generate a random IP, check if it is used and allocate it? The usage check could maybe even be a simple exception handler around the allocation. If it is in use we just generate a new one.

This has two benefits:
* We do not need to check what the last ip we used was and can just rely on the fact that the odds of duplicated ips are very low.
* Will therefore speed up the allocate next
* Will spread the used ips across the whole /64 and therefore not allow attackers to "guess" where the targets are. Otherwise an attacker would not need to scan the whole /64, but could stop scanning after e.g. 100 misses. Ok, this is just security through obscurity, but it will allow a SIEM to allert or applications like failtoban to blacklist the sourceip.

**Comment by neolit123** [comment]
xref https://github.com/kubernetes/kubernetes/pull/91875

**Comment by danwinship** [comment]
Another possibility would be to say that with IPv6, when the service CIDR is a /64:
- manually-allocated IPs _must_ be in the bottom /112, and are tracked to prevent duplicates using the same mechanism we use today. eg, given a service CIDR of `fd02::/64`, manually allocated IPs would be required to be in the range `fd02::1` to `fd02::ffff`. Trying to manually allocate an IP outside that range is an error, just like trying to claim an IP that's already in use.
- automatically-allocated IPs will use the remaining `2^64 - 2^16` addresses, and will be random and untracked. (That is, we will randomly pick a number from `0` to `2^64 - 1`, but if it's less than `2^16` we'll throw it away and try again.)

This deterministically ensures that no manually-allocated service IPs will conflict with any other manually- or automatically-allocated service IPs, and probabilistically ensures that no automatically-allocated service IPs will conflict with other automatically-allocated service IPs.

(Though this means you can't delete a Service that had an automatically-allocated IP and then re-create it with the same IP it had before. Is that a thing people care about doing?)

**Comment by agowa338** [comment]
I think it's good to say that if one did not specify an IP address explicitly it's reasonable to threat it as ephemeral. If you care, either use dns or assign a fixed one.
The suggestion with splitting the ranges and using the lower /112 for explicit assigned addresses looks to me as the cleanest so far.
The only thing that is still left is that people could want to "code meaning" into the address, like e.g. Facebook did with 2a03:2880:f11c:8083:**face:b00c**:0:25de.

**Comment by lavalamp** [comment]
That works for /64 but not for e.g. /80. Please see my earlier comment re: birthday approximations: https://github.com/kubernetes/kubernetes/pull/90115#issuecomment-616670315

If you want a two phase thing, when does it switch from that to deterministically tracking everything?

**Comment by agowa338** [comment]
Two things,
1. Why should one use /80? It's IPv6 we're talking about, a /64 is conceptionally one host. Therefore everything smaller with a prefix > /64 is just not typical. To address bad assignment policies of some cloud providers assuming to have a /64 assigned to the customer instead of the RIPE recommendation of /56 (for residential and /48 for business) should be fine. Therefore using /80 as the minimum should be a reasonable assumption. I cannot see any reason why someone would assign such cursed prefixes, please let me know if there is one. Everything smaller than /64 is prone to cause routing overhead or other kinds of issues anyway... I'd like to reference https://tools.ietf.org/html/rfc7421 for further details on this split as well as https://www.ripe.net/publications/docs/ripe-738 for the assignment guideline.
2. Because we separate the lower /120, we could just fall back on using one of the tracked ones if we cannot find a random one after a few guesses. And than we could just add it as a static address ourselves to the definition.

**Comment by danwinship** [comment]
As in the original proposal, you can either use a `/108` or greater for backward-compatibility, in which case it would use the old code paths exactly like it always did (ie, fully deterministic), or you can *exactly* `/64`, in which case it would behave as described above. You cannot use anything between `/65` and `/107`.

**Comment by aojea** [comment]
> Here's an idea:
> 
> * Make the bit array a fixed size.
> * If the CIDR is larger than that fixed size, set the bit (IP number) % (bit array size).
> * This means IPs with the same number % size will collide. That will be confusing for users, but a good error message will help.
> * A tactically chosen table size (prime, at least relatively with 2 and 10) may make that message rarer.

I tried all the approaches suggested in this thread, and I have something working with this one suggested by @lavalamp that seems feasible because the risk of collision is high, and only may happen in specific scenarios.

https://github.com/kubernetes/kubernetes/compare/master...aojea:service-network-ipv6

I'll write a KEP as follow up,
 thanks to all the contributions

**Comment by lavalamp** [comment]
@aojea 

> I tried all the approaches suggested in this thread, and I have something working ...

I just saw this PR closed for age-- is your prototype discussed elsewhere?

**Comment by aojea** [comment]
> I just saw this PR closed for age-- is your prototype discussed elsewhere?

@lavalamp the conversation continued here https://github.com/kubernetes/enhancements/pull/1881#issuecomment-672181143 but I didn't follow up because with dual stack, there were a lot of changes impacting the allocators, and  it was a very bad timing.

However, I really likek your idea and I want to explore it during this release
> Right, this is possible too, PVC/PV like. And is actually a much better model in that it gets us away from hidden, user-untouchable resources.

Do you have some link or pointers for the PVC/PV model, I've tried to find some references in the code, but I didn' t know where to start ...

**Comment by lavalamp** [comment]
Re PV/PVC: 

Honestly, I'd start with the user documentation: https://kubernetes.io/docs/concepts/storage/persistent-volumes/

I don't think the controller code is super easy to read, but it's here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L59

**Comment by aojea** [comment]
> PV/PVC:
> 
> Honestly, I'd start with the user documentation: https://kubernetes.io/docs/concepts/storage/persistent-volumes/
> 
> I don't think the controller code is super easy to read, but it's here: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L59

I was looking into this during the weekend and I think that we can do something similar but much simpler, let me a do a PoC to verify my hypothesis and come back to you.

**Comment by aojea** [comment]
I was able to create a prototype with webhooks for allocation, and a controller for recycling the deleted IPs.
I´ve commented this in the last sig-network meeting with Tim Hockin and he made a good point, this doesn´t roll back if the create/update request fails :/

He´s working right now in refactoring the service API https://github.com/kubernetes/kubernetes/pull/96684 , will try to check this again once he finish the refactoring

**Comment by lavalamp** [comment]
Yeah, you have to have a reconciliation loop, just as with the quota
controller.

@deads2k this is a spot where an "on failure" webhook might make sense.
Also cc @liggitt

**Comment by aojea** [comment]
@lavalamp PTAL https://docs.google.com/document/d/1CJ9i6u5nwTR6snW6LFO_aUwQoXFWUNz3s8soh5RiFzs/edit#heading=h.70rkmfl0pgdi
I pinged you in the original kep thread too

**Pull request closed without merging**